### PR TITLE
chore: Refactored ProvisionerManager. Migrated provisioners

### PR DESCRIPTION
**Pull request by nsainaney**
Depends on https://github.com/c6o/kubeclient/pull/20

### CODE REVIEW

File: packages/common/src/helpers/deployment.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { Cluster } from '@c6o/kubeclient-contracts'
+import { Deployment } from '@c6o/kubeclient-resources/apps/v1'
+import { DeploymentHelper as DeploymentHelperContract } from '@provisioner/contracts'
+
+export class DeploymentHelper<T extends Deployment = Deployment> extends DeploymentHelperContract<T> {
```
Comment (robblovell) [inline packages/common/src/helpers/deployment.ts:5]: Placeholder question: Is there a way to name "DeploymentHelper" to "Deployment".

File: packages/appengine/src/ui/views/base.ts [hunk 2]
```diff
@@ -42,9 +42,6 @@ export class BaseViewSettings extends AppEngineBaseView implements StoreFlowStep
     }
 
     handleLayout(items, type) {
-
-        this.state.startTimer('ui-configs-handleLayout')
```
Comment (robblovell) [inline packages/appengine/src/ui/views/base.ts:46]: Was the timer just for debugging purposes, or has this functionality been moved?

File: packages/c6o-system/src/contracts.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { ProvisionerBase, AppDocument } from '@provisioner/contracts'
+
+export interface SystemProvisioner extends ProvisionerBase {
+
+    postCreateApp(appSpec: AppDocument): Promise<void>
+    postUpdateApp(appSpec: AppDocument): Promise<void>
+    postRemoveApp(appSpec: AppDocument): Promise<void>
+    getSystemFQDN(): Promise<string>
```
Comment (robblovell) [inline packages/c6o-system/src/contracts.ts:8]: Does this belong here?

File: packages/common/src/helpers/update.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import { Cluster, PatchOps, Resource } from '@c6o/kubeclient-contracts'
+import * as pointer from 'jsonpointer'
+import createDebug from 'debug'
+
+const debug = createDebug('provisioner:c6o-system:updateSystem:')
+
+export const updateImageTag = async (cluster: Cluster, document: Resource, tag: string, path: string) => {
```
Comment (robblovell) [inline packages/common/src/helpers/update.ts:7]: Should the name of this file be more precisely named? something like updateImageTag.ts,  imageTag.ts, or image.ts?

File: packages/contracts/src/generators/namespace.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { Namespace } from '@c6o/kubeclient-resources/core/v1'
+import { setManagedBy } from '../markers'
+
+export const toNamespace = (namespace: string | Namespace): Namespace => {
```
Comment (robblovell) [inline packages/contracts/src/generators/namespace.ts]: I assume this is moving to a helper.from static method.

File: packages/contracts/src/resolver.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import { AppDocument } from './app'
+import { ProvisionerBase } from './provisioner'
+
+export type optionFunctionType = (string, description: string, autocomplete?: ReadonlyArray<string>) => void
+
+export type actionType = 'create' | 'update' | 'remove'
+export type stageType = 'load' | 'inquire' | 'validate' | 'apply' | 'done'
+export type transactionStateType = 'pending' | 'completed' | 'error'
+
+/*export interface ProvisionerManager extends EventEmitter {
```
Comment (robblovell) [inline packages/contracts/src/resolver.ts]: Update the comment here and provide more information about what "Resolver" does.

File: packages/contracts/src/provisioner.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { Cluster, Status } from '@c6o/kubeclient-contracts'
+import { AppDocument } from './app'
+import { Resolver,  } from './resolver'
+
+export interface IngressParameters {
+    ip?: string
+    hostname?: string
+}
+export interface ProvisionerBase {
```
Comment (robblovell) [inline packages/contracts/src/provisioner.ts]: Need a comment here describing what provisioner base does. Essentially what is a provisioner?

File: packages/grafana/src/contracts.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { ConfigMap } from '@c6o/kubeclient-resources/core/v1';
+import { ProvisionerBase } from '@provisioner/contracts'
+
+export interface GrafanaProvisioner extends ProvisionerBase {
```
Comment (robblovell) [inline packages/grafana/src/contracts.ts:4]: Add a comment here about what Grafana is and where and why it's used.

File: packages/contracts/src/markers.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { Resource, setLabel, setAnnotation } from '@c6o/kubeclient-contracts'
+
+const keyManagedBy = 'app.kubernetes.io/managed-by'
+
+// Labels - searchable
```
Comment (robblovell) [inline packages/contracts/src/markers.ts]: Better comment about what a marker is and how it should be used.

File: packages/common/src/helpers/deployment.ts (hunk 1 above)
Comment (robblovell) [inline packages/common/src/helpers/deployment.ts:5]: Need a readme in the helpers directory describing what a helper is.

File: packages/istio/src/contracts.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { Result } from '@c6o/kubeclient-contracts'
+import { ProvisionerBase, IngressParameters, AppDocument } from '@provisioner/contracts'
+
+export interface IstioProvisioner extends ProvisionerBase {
```
Comment (robblovell) [inline packages/istio/src/contracts.ts:4]: Need a comment here about what Istio is and how it is used.

File: packages/mongo/src/mixins/createInquire.ts [hunk 11]
```diff
@@ -15,14 +17,14 @@ export const createInquireMixin = (base: baseProvisionerType) => class extends b
         return this.spec.secretKeyRef || args['secret-key']
     }
 
-    async inquire(args) {
```
Comment (robblovell) [inline packages/mongo/src/mixins/createInquire.ts:18]: Is merging Inquire and CreateInquire okay? Should the collecting of information be separated from the action taken on the information? Weren't there provisioners that needed these separate?

File: packages/common/src/helpers/deployment.ts (hunk 1 above)
Comment (nsainaney) [reply packages/common/src/helpers/deployment.ts:5]: We have discussed this - please mark resolved

File: packages/appengine/src/ui/views/base.ts (hunk 2 above)
Comment (nsainaney) [reply packages/appengine/src/ui/views/base.ts:46]: Robb wanted to meter everything but this stuff was removed (this code was a holdover)

File: packages/c6o-system/src/contracts.ts (hunk 3 above)
Comment (nsainaney) [reply packages/c6o-system/src/contracts.ts:8]: WIP - will probably get cleaned up

File: packages/contracts/src/generators/namespace.ts (hunk 5 above)
Comment (nsainaney) [reply packages/contracts/src/generators/namespace.ts]: WIP - will move

File: packages/common/src/helpers/service.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import { Cluster, ResourceHelper } from '@c6o/kubeclient-contracts'
+import { Service } from '@c6o/kubeclient-resources/core/v1'
+import { IngressParameters } from '@provisioner/contracts'
+//import { ServiceHelper as ServiceHelperContract } from '@provisioner/contracts'
```
Comment (gburnard) [inline packages/common/src/helpers/service.ts]: Should this be removed?

File: packages/istio/src/mixins/api/gateway.ts [hunk 13]
```diff
@@ -38,30 +39,25 @@ export const gatewayApiMixin = (base: baseProvisionerType) => class extends base
...
-            kind: 'Service',
-            metadata: {
-                namespace: 'istio-system',
-                labels: {
-                    istio: 'ingressgateway'
-                }
-            }
-        })
+        return await ServiceHelper
+            .from('istio-system')
```
Comment (gburnard) [inline packages/istio/src/mixins/api/gateway.ts:59]: Should we set the namespace globally in a const?

**Review by gburnard — COMMENTED** [review]
Looks good.  Added a few comments.  I don't see any unit tests, though.

**Pull request merged**